ckanext/search/filters.py

```diff
@@ -4,51 +4,189 @@
+        return {"eq": value}
+
+    elif isinstance(value, dict):
+
+        if len(value.keys()) > 1:
+            # Combine dict filters with $and
+            members = [{field_name: {k: v}} for k, v in value.items()]
+            return {"$and": _process_filter_operator_members(members)}
+
+        else:
+            # Just return the filter
+            # TODO: check op and value format
+            return value
+
+    elif isinstance(value, list):
+        # TODO: fail if lists
+
+        field_ops = [x for x in value if isinstance(x, dict)]
+        non_field_ops = [x for x in value if x not in field_ops]
+
+        if not field_ops:
+            return {"in": value}
+
+        members = []
+        for field_op in field_ops:
+            members.append(
+                {"$and": [{field_name: {k: v}} for k, v in field_op.items()]}
+            )
+
+        members.append({field_name: {"in": non_field_ops}})
+
+        return {"$or": members}
+
+
+def query_filters_validator(
+    value: Optional[Union[Dict[str, Any], List[Dict[str, Any]]]],
+) -> Optional[FilterOp]:
 
     if not value:
-        return
+        return None
 
     # Check structure
     if not _is_dict_or_list_of_dicts(value):
         raise Invalid("Filters must be defined as a dict or a list of dicts")
 
+    filters = {}
+
     # Normalize to dict
     if isinstance(value, list):
-        filters = {"$and": value}
+        filters = {"$or": value}
+
     else:
-        filters = value
-
-    # Check top operators
-    top_operators = [
-        k for k in filters.keys() if k.startswith("$") and not k.startswith("$$")
-    ]
-    if unknown_top_operators := [k for k in top_operators if k not in TOP_OPERATORS]:
-        raise Invalid(
-            f"Unknown operators (must be one of $or, $and): {unknown_top_operators}"
-        )
 
-    # Top operator values must be lists of dicts or dicts
-    for top_operator in top_operators:
-        if not _is_dict_or_list_of_dicts(filters[top_operator]):
-            raise Invalid(
-                "Top level operations must be defined as a dict or a list of dicts"
-            )
+        if len(value.keys()) == 1:
+
+            key = list(value.keys())[0]
+
+            if key.startswith("$") and not key.startswith("$$"):
+                _check_filter_operator(key, value[key])
+                filters[key] = _process_filter_operator_members(value[key])
+            else:
+                filters[key] = _process_field_operator(key, value[key])
+
+        else:
+            # Move individual filters to $and
+            for k, v in value.items():
+
+                if "$and" not in filters:
+                    filters["$and"] = []
+
+                filters["$and"].append({k: v})
 
-    # Individual filters should
+    for op in ("$and", "$or"):
+        if op in filters:
+            filters[op] = _process_filter_operator_members(filters[op])
 
+    # Translate output dict to immutable FilterOp
+    out = _to_filter_op(filters)
```

Comment:
Is there a reason to do this as a separate recursive step instead of as part of the processing/validation?